numpy/lib/function_base.py

```diff
@@ -175,6 +198,93 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
             raise AttributeError(
                 'max must be larger than min in range parameter.')
 
+    from numpy.compat.py3k import basestring
+    if isinstance(bins, basestring):
+        # Automatically determine bins. Defined as functions here for easy refactoring
+        # Can easily be moved into relevant if functions or separate file
+
+        def sturges(x):
+            """
+            Sturges Estimator
+            A very simplistic estimator based on the assumption of normality of the data
+            Poor performance for non-normal data, especially obvious for large X.
+            Depends only on size of the data
+            """
+            return np.ceil(np.log2(x.size)) + 1
+
+        def rice(x):
+            """
+            Rice Estimator
+            Another simple estimator, with no normality assumption.
+            It has better performance for large data, but tends to be too large
+            The number of bins is proportional to the cube root of data size (asymptotically optimal)
+            Depends only on size of the data
+            """
+            return np.ceil(2 * x.size ** (1.0/3))
+
+        def scott(x):
+            """
+            Scott Estimator
+            The binwidth is proportional to the standard deviation of the data and
+            inversely proportional to the cube root of data size (asymptotically optimal)
+
+            """
+            h = 3.5 * x.std() * x.size **(-1.0/3)
+            if h>0:
+                return np.ceil(x.ptp()/h)
+            return 1
+
+        def mad(data, axis=None):
+            """
+            Mean Absolute Deviation - a simple well defined function
+            Simplest implementation, even length case left to default behaviour
+            This is used by the Freedman Diaconis estimator when the IQR does not exist
+            """
+            return np.median(np.absolute(data - np.median(data, axis)), axis)
+
+        def FD(x):
+            """
+            Freedman Diaconis rule using Inter Quartile Range (IQR) for binwidth
+            Considered a variation of the Scott rule with more robustness as the IQR
+            is less affected by outliers than the standard deviation.
+
+            If the IQR is 0, we return the median absolute deviation as defined above, else 1
+            Binwidth is inversely proportional to the cube root of data size (asymptotically optimal)
+            """
+            iqr = np.subtract(*np.percentile(x, [75, 25]))
+
+            if iqr ==0: #unlikely
+                iqr = mad(x) # replace with something useful
+
+            h = (2 * iqr * x.size**(-1.0/3)) # binwidth
+
+            if iqr > 0:
+                return np.ceil(x.ptp()/h)
+
+            return 1 #all else fails
+
+        def default(x):
```

Comment:
Better to name this function ``auto``, because that's what the string is called and because it's not actually the default.